File: game/atc-core/server/medical.lua
-- ATC Medical Bridge — server-authoritative EMS, trauma & medical operations
-- All principal IDs resolved server-side; no client value trusted.

local API_BASE  = ATC.Config.ApiBase  or 'http://localhost:3000'
local API_TOKEN = ATC.Config.ApiToken or ''

local function apiPost(path, body)
  local payload = json.encode(body)
  local result = PerformHttpRequest(API_BASE .. path, function() end, 'POST', payload, {
    ['Content-Type']  = 'application/json',
    ['Authorization'] = 'Bearer ' .. API_TOKEN,
  })
  return result
end

local function apiGet(path)
  local result = PerformHttpRequest(API_BASE .. path, function() end, 'GET', '', {
    ['Authorization'] = 'Bearer ' .. API_TOKEN,
  })
  return result
end

local function apiPatch(path, body)
  local payload = json.encode(body)
  local result = PerformHttpRequest(API_BASE .. path, function() end, 'PATCH', payload, {
    ['Content-Type']  = 'application/json',
    ['Authorization'] = 'Bearer ' .. API_TOKEN,
  })
  return result
end

-- ── Public SDK ───────────────────────────────────────────────────────────────

ATC.Medical = {}

--- Record an injury for a character.
--- @param source       number  FiveM player server id of the recording officer/medic
--- @param characterId  string
--- @param region       string  body region
--- @param severity     string  'minor'|'moderate'|'critical'|'fatal'
--- @param description  string
--- @param incidentId   string|nil  optional linked incident
--- @return table|nil
function ATC.Medical.RecordInjury(source, characterId, region, severity, description, incidentId)
  local principalId = ATC.Accounts.GetPrincipalId(source)
  if not principalId then return nil end
  return apiPost('/api/v1/medical/injuries', {
    characterId           = characterId,
    recordedByPrincipalId = principalId,
    region                = region,
    severity              = severity,
    description           = description,
    incidentId            = incidentId,
  })
end

--- Get the current trauma state for a character.
--- @param characterId string
--- @return table|nil
function ATC.Medical.GetTrauma(characterId)
  return apiGet('/api/v1/medical/trauma/' .. characterId)
end

--- Transition a character's trauma state.
--- @param source       number   FiveM player server id
--- @param characterId  string
--- @param newState     string   trauma state
--- @param notes        string|nil
--- @return table|nil
function ATC.Medical.UpdateTrauma(source, characterId, newState, notes)
  local principalId = ATC.Accounts.GetPrincipalId(source)
  if not principalId then return nil end
  return apiPatch('/api/v1/medical/trauma/' .. characterId, {
    newState             = newState,
    updatedByPrincipalId = principalId,
    notes                = notes,
  })
end

--- Revive a deceased character. Requires ems.revive capability on the caller.
--- @param source       number   FiveM player server id of the medic
--- @param characterId  string
--- @param incidentId   string|nil
--- @param notes        string|nil
--- @return table|nil
function ATC.Medical.RevivePatient(source, characterId, incidentId, notes)
  local principalId = ATC.Accounts.GetPrincipalId(source)
  if not principalId then return nil end
  return apiPost('/api/v1/medical/revive/' .. characterId, {
    revivedByPrincipalId = principalId,
    incidentId           = incidentId,
    notes                = notes,
  })
end

--- Apply a treatment to a character.
--- @param source       number   FiveM player server id
--- @param characterId  string
--- @param treatmentType string  e.g. 'bandage', 'cpr', 'defibrillator'
--- @param opts         table    optional: incidentId, itemId, notes, previousTrauma, resultingTrauma
--- @return table|nil
function ATC.Medical.ApplyTreatment(source, characterId, treatmentType, opts)
  local principalId = ATC.Accounts.GetPrincipalId(source)
  if not principalId then return nil end
  opts = opts or {}
  return apiPost('/api/v1/medical/treatments', {
    characterId          = characterId,
    appliedByPrincipalId = principalId,
    type                 = treatmentType,
    incidentId           = opts.incidentId,
    itemId               = opts.itemId,
    notes                = opts.notes,
    previousTrauma       = opts.previousTrauma,
    resultingTrauma      = opts.resultingTrauma,
  })
end

--- Create a medical report.
--- @param source       number   FiveM player server id
--- @param characterId  string
--- @param diagnosis    string
--- @param opts         table    optional: incidentId, arrestId, notes, injuryIds, treatmentIds, vitalsSnapshot
--- @return table|nil
function ATC.Medical.CreateReport(source, characterId, diagnosis, opts)
  local principalId = ATC.Accounts.GetPrincipalId(source)
  if not principalId then return nil end
  opts = opts or {}
  return apiPost('/api/v1/medical/reports', {
    characterId          = characterId,
    createdByPrincipalId = principalId,
    diagnosis            = diagnosis,
    incidentId           = opts.incidentId,
    arrestId             = opts.arrestId,
    notes                = opts.notes,
    injuryIds            = opts.injuryIds,
    treatmentIds         = opts.treatmentIds,
    vitalsSnapshot       = opts.vitalsSnapshot,
  })
end

--- Admit a character to hospital.
--- @param source       number   FiveM player server id
--- @param characterId  string
--- @param opts         table    optional: facilityId, incidentId, notes
--- @return table|nil
function ATC.Medical.AdmitToHospital(source, characterId, opts)
  local principalId = ATC.Accounts.GetPrincipalId(source)
  if not principalId then return nil end
  opts = opts or {}
  return apiPost('/api/v1/medical/hospital/admit', {
    characterId          = characterId,
    admittedByPrincipalId = principalId,
    facilityId           = opts.facilityId,
    incidentId           = opts.incidentId,
    notes                = opts.notes,
  })
end

--- Get active hospital record for a character.
--- @param characterId string
--- @return table|nil
function ATC.Medical.GetHospitalRecord(characterId)
  return apiGet('/api/v1/medical/hospital/character/' .. characterId)
end

--- Update hospital status (e.g. admitted → icu → surgery → discharged).
--- @param source     number   FiveM player server id
--- @param recordId   string   hospital record id
--- @param newStatus  string
--- @param notes      string|nil
--- @return table|nil
function ATC.Medical.UpdateHospitalStatus(source, recordId, newStatus, notes)
  local principalId = ATC.Accounts.GetPrincipalId(source)
  if not principalId then return nil end
  return apiPatch('/api/v1/medical/hospital/' .. recordId .. '/status', {
    newStatus            = newStatus,
    updatedByPrincipalId = principalId,
    notes                = notes,
  })
end

File: packages/shared-types/src/medical.ts
// ── Primitives ─────────────────────────────────────────────────────────────────

export type AtcMedicalSeverity = 'minor' | 'moderate' | 'critical' | 'fatal'

export type AtcBodyRegion =
  | 'head'
  | 'chest'
  | 'abdomen'
  | 'left_arm'
  | 'right_arm'
  | 'left_leg'
  | 'right_leg'
  | 'spine'

export type AtcTraumaState =
  | 'stable'
  | 'bleeding'
  | 'unconscious'
  | 'cardiac_arrest'
  | 'fractured'
  | 'pain_shock'
  | 'stabilized'
  | 'deceased'

export type AtcHospitalStatus = 'admitted' | 'icu' | 'surgery' | 'discharged' | 'deceased'

export type AtcTreatmentType =
  | 'bandage'
  | 'defibrillator'
  | 'medication'
  | 'splint'
  | 'tourniquet'
  | 'cpr'
  | 'revive'
  | 'stabilize'
  | 'transfer'
  | 'other'

// ── Injury Record ─────────────────────────────────────────────────────────────

export interface AtcInjuryRecord {
  id: string
  characterId: string
  agencyId: string | null
  incidentId: string | null
  recordedByPrincipalId: string
  region: AtcBodyRegion
  severity: AtcMedicalSeverity
  description: string
  metadata: Record<string, unknown>
  createdAt: Date
  updatedAt: Date
}

// ── Trauma State ──────────────────────────────────────────────────────────────

export interface AtcTraumaRecord {
  id: string
  characterId: string
  state: AtcTraumaState
  previousState: AtcTraumaState | null
  updatedByPrincipalId: string
  notes: string | null
  stateChangedAt: Date
  createdAt: Date
  updatedAt: Date
}

// ── Treatment Record (append-only) ───────────────────────────────────────────

export interface AtcTreatmentRecord {
  id: string
  characterId: string
  appliedByPrincipalId: string
  incidentId: string | null
  type: AtcTreatmentType
  itemId: string | null
  notes: string | null
  previousTrauma: AtcTraumaState | null
  resultingTrauma: AtcTraumaState | null
  metadata: Record<string, unknown>
  appliedAt: Date
}

// ── Medical Report ────────────────────────────────────────────────────────────

export interface AtcMedicalReport {
  id: string
  characterId: string
  createdByPrincipalId: string
  incidentId: string | null
  arrestId: string | null
  diagnosis: string
  notes: string
  injuryIds: string[]
  treatmentIds: string[]
  vitalsSnapshot: Record<string, unknown> | null
  closedAt: Date | null
  closedByPrincipalId: string | null
  createdAt: Date
  updatedAt: Date
}

// ── Hospital Record ───────────────────────────────────────────────────────────

export interface AtcHospitalRecord {
  id: string
  characterId: string
  admittedByPrincipalId: string
  status: AtcHospitalStatus
  facilityId: string | null
  incidentId: string | null
  notes: string | null
  admittedAt: Date
  statusChangedAt: Date
  dischargedAt: Date | null
  updatedAt: Date
}

// ── Revive Request ────────────────────────────────────────────────────────────

export interface AtcReviveRequest {
  characterId: string
  revivedByPrincipalId: string
  incidentId: string | null
  notes: string | null
}

// ── Events ────────────────────────────────────────────────────────────────────

export const ATC_MEDICAL_EVENTS = {
  INJURY_RECORDED:        'atc:medical:injury:recorded',
  TRAUMA_ESCALATED:       'atc:medical:trauma:escalated',
  PLAYER_REVIVED:         'atc:medical:player:revived',
  TREATMENT_APPLIED:      'atc:medical:treatment:applied',
  MEDICAL_REPORT_CREATED: 'atc:medical:report:created',
  PATIENT_STABILIZED:     'atc:medical:patient:stabilized',
  PATIENT_DECEASED:       'atc:medical:patient:deceased',
} as const

export type AtcMedicalEventName = typeof ATC_MEDICAL_EVENTS[keyof typeof ATC_MEDICAL_EVENTS]
